ashley/src/main/kotlin/ktx/ashley/properties.kt

```diff
@@ -74,3 +74,38 @@ inline fun <reified T : Component> optionalPropertyFor(
   mapper: ComponentMapper<T> = mapperFor()
 ): OptionalComponentDelegate<T> =
   OptionalComponentDelegate(mapper, T::class.java)
+
+/**
+ * Property delegate for an [Entity] wrapping around a [ComponentMapper].
+ * Allows checking the presence of a component on an Entity.
+ * Allows assigning and removing a component from an Entity with a Boolean value
+ */
+
+class TagDelegate <T : Component> (val mapper: ComponentMapper<T>, private val componentClass: Class<T>) {
+
+  operator fun getValue(thisRef: Entity, property: KProperty<*>): Boolean =
+    mapper.has(thisRef)
+
+  operator fun setValue(thisRef: Entity, property: KProperty<*>, value: Boolean) {
+    if (value) {
+      thisRef.add(componentClass.getConstructor().newInstance())
```

Comment:
Using reflection is tricky. This will result in runtime exceptions if there is no no-arg constructor (which isn't a good use case for the `TagDelegate`, but still). In general, it would be better to provide either a singleton instance for immutable components or a component provider lambda if a singleton is not acceptable.

Reply:
Ashley.Ktx already have functionalities that require to have a no-arg constructor for Components and throw exception if not  ( The ``` engine.add { entity { with<Component> } } }``` does ), so I don't think it would be that bad.
I'd like to keep a ```tagFor``` function that works just with the type supplied, it's just so elegant. 
Maybe we can make the function tagFor throw an Exception if there is no no-arg constructor and no value supplied? 
The delegate tagFor could have a parameter that is the default value of the Component.
The tagFor function could have an optional argument that is the default value of the Component, and provide the value of the no-arg constructor by default, so the reflection operation only happens once at the creation of the property.

Reply:
Reflection should be avoided whenever possible. It's the most error-prone API on platforms such as iOS or experimental web, since they basically have to emulate JVM reflection. The performance might also vary from platform to platform - it's [pretty OK](https://stackoverflow.com/questions/435553/java-reflection-performance) on desktop (twice as slow, still kinda fast), but it might be worse on mobiles.

It's true that `ktx-ashley` provides functions such as `with` that rely on reflection but it also has a way to avoid them entirely with utilities such as the `+=` operator. For example, this syntax should be preferred for performance-critical applications:

```kotlin
  val entity = Entity()
  entity += Transform(x=4f, y=2f)

  // With engine:
  val engine = Engine()
  engine.entity { 
    entity += Transform(x=4f, y=2f)
  }
```

It's OK to provide a reflection-based solution to speed things up, but there should be an option to avoid it entirely.

Reply:
I think it's nice for prototyping, but we can mention in the doc that it is faster to provide a default Component value. In the last commit, the reflection operation happens only once per property declared, so It's not that great of deal.